src/peer_transport/transport.ts

```diff
@@ -0,0 +1,154 @@
+    })
+  }
+
+  async stop () {
+    await this.components.registrar.unhandle(PROTOCOL)
+  }
+
+  createListener (options: CreateListenerOptions): Listener {
+    return new WebRTCPeerListener(this.components)
+  }
+
+  get [Symbol.toStringTag] (): string {
+    return '@libp2p/webrtc-direct'
+  }
+
+  get [symbol] (): true {
+    return true
+  }
+
+  filter (multiaddrs: Multiaddr[]): Multiaddr[] {
+    return multiaddrs.filter((ma) => {
+      const codes = ma.protoCodes()
+      return codes.includes(CODE)
+    })
+  }
+
+  /*
+   * dial connects to a remote via the circuit relay or any other protocol
+   * and proceeds to upgrade to a webrtc connection.
+   * multiaddr of the form: <multiaddr>/webrtc-direct/p2p/<destination-peer>
+   * For a circuit relay, this will be of the form
+   * <relay address>/p2p/<relay-peer>/p2p-circuit/webrtc-direct/p2p/<destination-peer>
+  */
+  async dial (ma: Multiaddr, options: DialOptions): Promise<Connection> {
+    log.trace('dialing address: ', ma)
+    const addrs = ma.toString().split(TRANSPORT)
+    if (addrs.length !== 2) {
+      // TODO(ckousik): Change to errCode
+      throw new Error('invalid multiaddr')
+    }
+    // look for remote peerId
+    const remoteAddr = multiaddr(addrs[0])
+    const destination = multiaddr(addrs[1])
+
+    const destinationIdString = destination.getPeerId()
+    if (destinationIdString == null) {
+      // TODO(ckousik): Change to errCode
+      throw new Error('bad destination')
+    }
+
+    const controller = new AbortController()
+    if (options.signal == null) {
+      options.signal = controller.signal
+    }
+
+    const connection = await this.components.transportManager.dial(remoteAddr)
+
+    const rawStream = await connection.newStream([PROTOCOL], options)
+
+    const [pc, muxerFactory] = await connect({
+      stream: rawStream,
+      rtcConfiguration: this.init.rtcConfiguration,
+      signal: options.signal
+    })
+
+    rawStream.close()
+    return await options.upgrader.upgradeOutbound(
+      new WebRTCMultiaddrConnection({
+        peerConnection: pc,
+        timeline: { open: (new Date()).getTime() },
+        remoteAddr: connection.remoteAddr
+      }),
+      {
+        skipProtection: true,
+        skipEncryption: true,
+        muxerFactory
+      }
+    )
+  }
+
+  async _onProtocol ({ connection, stream }: IncomingStreamData) {
+    const [pc, muxerFactory] = await handleIncomingStream({
+      rtcConfiguration: this.init.rtcConfiguration,
+      connection,
+      stream
+    })
+    const conn = await this.components.upgrader.upgradeInbound(new WebRTCMultiaddrConnection({
+      peerConnection: pc,
+      timeline: { open: (new Date()).getTime() },
+      remoteAddr: connection.remoteAddr
+    }), {
+      skipEncryption: true,
+      skipProtection: true,
+      muxerFactory
+    })
+    stream.close()
```

Comment:
Can we move this close to above the `const conn =` line so it's similar to the `dial` method, please?